official/utils/arg_parsers/parsers_test.py

```diff
@@ -0,0 +1,77 @@
+# Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+
+import argparse
+import unittest
+
+
+import official.utils.arg_parsers
+
+
+class TestParser(argparse.ArgumentParser):
+  """Class to test canned parser functionality."""
+
+  def __init__(self):
+    super(TestParser, self).__init__(parents=[
+        official.utils.arg_parsers.LocationParser(data_dir=True,
+                                                  model_dir=True),
+        official.utils.arg_parsers.DeviceParser(multi_gpu=True),
+        official.utils.arg_parsers.SupervisedParser(
+            train_epochs=True, epochs_per_eval=True, batch_size=True),
+        official.utils.arg_parsers.PerformanceParser(
+            num_parallel_calls=True, inter_op=True, intra_op=True),
+        official.utils.arg_parsers.DummyParser(use_synthetic_data=True),
+        official.utils.arg_parsers.ImageModelParser(data_format=True),
+    ])
+
+
+class BaseTester(unittest.TestCase):
+
+  def test_default_setting(self):
+    """Test to ensure fields exist and defaults can be set.
+    """
+
+    defaults = dict(
+        data_dir="dfgasf",
+        model_dir="dfsdkjgbs",
+        train_epochs=534,
+        epochs_per_eval=15,
+        batch_size=256,
+        num_parallel_calls=18,
+        inter_op_parallelism_threads=5,
+        intra_op_parallelism_thread=10,
+        data_format="channels_first"
+    )
+
+    parser = TestParser()
+    parser.set_defaults(**defaults)
+
+    namespace_vars = vars(parser.parse_args([]))
+    for key, value in defaults.items():
+      assert namespace_vars[key] == value
+
+  def test_booleans(self):
+    """Test to ensure boolean flags trigger as expected.
+    """
+
+    parser = TestParser()
+    namespace = parser.parse_args(["--multi_gpu", "--use_synthetic_data"])
+
+    assert namespace.multi_gpu
+    assert namespace.use_synthetic_data
+
```

Comment:
Do u also want to test the case that the flag value is specified while the arg parser is not turned on?

Reply:
This turns out to not be practical. When parse_args fails, argparse fails hard by calling sys.exit(2) rather than raising an exception. While I certainly could invoke subprocess and assert an exit code of 2, this seems like too much machinery for a test that is just intended to show off parse_args().